plugins/providers/gcs/provider.go

```diff
@@ -0,0 +1,253 @@
+package gcs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"cloud.google.com/go/iam"
+	"github.com/mitchellh/mapstructure"
+	"github.com/odpf/guardian/core/provider"
+	"github.com/odpf/guardian/domain"
+	"github.com/odpf/guardian/utils"
+)
+
+//go:generate mockery --name=Crypto --exported --with-expecter
+type Crypto interface {
+	domain.Crypto
+}
+type Provider struct {
+	provider.PermissionManager
+
+	typeName string
+	Clients  map[string]GCSClient
+	crypto   Crypto
+}
+
+func NewProvider(typeName string, crypto Crypto) *Provider {
+	return &Provider{
+		typeName: typeName,
+		Clients:  map[string]GCSClient{},
+		crypto:   crypto,
+	}
+}
+
+func (p *Provider) CreateConfig(pc *domain.ProviderConfig) error {
+	c := NewConfig(pc, p.crypto)
+
+	if err := c.ParseAndValidate(); err != nil {
+		return err
+	}
+
+	return c.EncryptCredentials()
+}
+
+func (p *Provider) GetResources(pc *domain.ProviderConfig) ([]*domain.Resource, error) {
+	var creds Credentials
+	if err := mapstructure.Decode(pc.Credentials, &creds); err != nil {
+		return nil, err
+	}
+
+	if err := creds.Decrypt(p.crypto); err != nil {
+		return nil, err
+	}
+
+	client, err := p.getGCSClient(creds)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceTypes := []string{}
+	for _, rc := range pc.Resources {
+		resourceTypes = append(resourceTypes, rc.Type)
+	}
+
+	var resources []*domain.Resource
+	projectID := strings.Replace(creds.ResourceName, "projects/", "", 1)
+	buckets, err := client.GetBuckets(context.TODO(), projectID)
+	if err != nil {
+		return nil, err
+	}
+	for _, b := range buckets {
+		bucketResource := b.toDomain()
+		bucketResource.ProviderType = pc.Type
+		bucketResource.ProviderURN = pc.URN
+		if utils.ContainsString(resourceTypes, ResourceTypeBucket) {
+			resources = append(resources, bucketResource)
+		}
+	}
+
+	return resources, nil
+}
+
+func (p *Provider) GetType() string {
+	return p.typeName
+}
+
+func (p *Provider) GrantAccess(pc *domain.ProviderConfig, a *domain.Appeal) error {
+	if err := validateProviderConfigAndAppealParams(pc, a); err != nil {
+		return fmt.Errorf("invalid provider/appeal config: %w", err)
+	}
+
+	permissions, err := getPermissions(pc.Resources, a)
+	if err != nil {
+		return fmt.Errorf("error in getting permissions: %w", err)
+	}
+
+	var creds Credentials
+	if err := mapstructure.Decode(pc.Credentials, &creds); err != nil {
+		return fmt.Errorf("error in decoding credentials%w", err)
+	}
+
+	if err := creds.Decrypt(p.crypto); err != nil {
+		return fmt.Errorf("error in decrypting credentials%w", err)
+	}
+
+	client, err := p.getGCSClient(creds)
+	if err != nil {
+		return fmt.Errorf("error in getting new client: %w", err)
+	}
+
+	identity := fmt.Sprintf("%s:%s", a.AccountType, a.AccountID) // identity is AccountType : AccountID, eg: "serviceAccount:test@email.com"
```

Comment:
you can put the comment above it to make it easier to read
```suggestion
        // identity is AccountType : AccountID, eg: "serviceAccount:test@email.com"
	identity := fmt.Sprintf("%s:%s", a.AccountType, a.AccountID) 
```

Reply:
done in https://github.com/odpf/guardian/pull/227/commits/575cacf66a8181ddf2953708d3c7d26f0b98eba5